test_runner/batch_others/test_wal_acceptor_async.py

```diff
@@ -456,3 +457,68 @@ def test_unavailability(neon_env_builder: NeonEnvBuilder):
     pg = env.postgres.create_start('test_safekeepers_unavailability')
 
     asyncio.run(run_unavailability(env, pg))
+
+
+@dataclass
+class RaceConditionTest:
+    iteration: int
+    is_stopped: bool
+
+
+# shut down random subset of safekeeper, sleep, wake them up, rinse, repeat
+async def xmas_garland(safekeepers: List[Safekeeper], data: RaceConditionTest):
+    while not data.is_stopped:
+        data.iteration += 1
+        victims = []
+        for sk in safekeepers:
+            if random.random() >= 0.5:
+                victims.append(sk)
+        log.info(
+            f'Iteration {data.iteration}: stopping {list(map(lambda sk: sk.id, victims))} safekeepers'
+        )
+        for v in victims:
+            v.stop()
+        await asyncio.sleep(1)
+        for v in victims:
+            v.start()
+        log.info(f'Iteration {data.iteration} finished')
+        await asyncio.sleep(1)
+
+
+async def run_race_conditions(env: NeonEnv, pg: Postgres):
+    conn = await pg.connect_async()
+    await conn.execute('CREATE TABLE t(key int primary key, value text)')
+
+    data = RaceConditionTest(0, False)
+    bg_xmas = asyncio.create_task(xmas_garland(env.safekeepers, data))
+
+    n_iterations = 5
+    n_queries = 200
+    expected_sum = 0
+
+    for i in range(1, n_iterations + 1):
+        # await for the end of the previous iteration
+        while data.iteration < i:
+            await asyncio.sleep(0.1)
+
+        for i in range((i - 1) * n_queries, i * n_queries):
+            await conn.execute(f"INSERT INTO t values ({i+1}, 'payload')")
+            expected_sum += i + 1
```

Comment:
This is a creative way to synchronize between the tasks, it took me a while to understand how it works. Is there a reason you need to wait for the previous iteration to finish, before you execute more INSERTs?

```suggestion
    while data.iteration < n_iterations:
            await asyncio.sleep(0.01)
            await conn.execute(f"INSERT INTO t values ({i+1}, 'payload')")
            expected_sum += i + 1
```

Reply:
No specific reason to wait for the previous iteration, just wanted to execute queries in different iterations. Your suggestion is much better for that, thanks!